pkg/virt-controller/watch/vm.go

```diff
@@ -2025,6 +1921,130 @@ func (c *VMController) processFailureCondition(vm *virtv1.VirtualMachine, vmi *v
+					clearChangeRequest = true
+				}
+			}
+		case virtv1.StartRequest:
+			// If the current VMI is running, then it has been started.
+			if vmi != nil {
+				log.Log.Object(vm).V(4).Infof("VMI exists. clearing start request")
+				clearChangeRequest = true
+			}
+		}
+	}
+
+	return clearChangeRequest
+}
+
+func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
+	if len(vm.Status.VolumeRequests) > 0 {
+		volumeMap := make(map[string]virtv1.Volume)
+		diskMap := make(map[string]virtv1.Disk)
+
+		for _, volume := range vm.Spec.Template.Spec.Volumes {
+			volumeMap[volume.Name] = volume
+		}
+		for _, disk := range vm.Spec.Template.Spec.Domain.Devices.Disks {
+			diskMap[disk.Name] = disk
+		}
+
+		tmpVolRequests := vm.Status.VolumeRequests[:0]
+		for _, request := range vm.Status.VolumeRequests {
+
+			var added bool
+			var volName string
+
+			removeRequest := false
+
+			if request.AddVolumeOptions != nil {
+				volName = request.AddVolumeOptions.Name
+				added = true
+			} else if request.RemoveVolumeOptions != nil {
+				volName = request.RemoveVolumeOptions.Name
+				added = false
+			}
+
+			_, volExists := volumeMap[volName]
+			_, diskExists := diskMap[volName]
+
+			if added && volExists && diskExists {
+				removeRequest = true
+			} else if !added && !volExists && !diskExists {
+				removeRequest = true
+			}
+
+			if !removeRequest {
+				tmpVolRequests = append(tmpVolRequests, request)
+			}
+		}
+		vm.Status.VolumeRequests = tmpVolRequests
+	}
+}
+
+func (c *VMController) handleScalingUpOrDown(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance, key string, dataVolumes []*cdiv1.DataVolume) (syncError, error) {
+	var syncErr syncError
+
+	// Scale up or down, if all expected creates and deletes were report by the listener
+	if c.needsSync(key) && vm.ObjectMeta.DeletionTimestamp == nil {
+		runStrategy, err := vm.RunStrategy()
+		if err != nil {
+			return nil, err
+		}
+
+		dataVolumesReady, err := c.handleDataVolumes(vm, dataVolumes)
+		if err != nil {
+			syncErr = &syncErrorImpl{fmt.Errorf("Error encountered while creating DataVolumes: %v", err), FailedCreateReason}
+		} else if dataVolumesReady || runStrategy == virtv1.RunStrategyHalted {
+			syncErr = c.startStop(vm, vmi)
+		} else {
+			log.Log.Object(vm).V(3).Infof("Waiting on DataVolumes to be ready. %d datavolumes found", len(dataVolumes))
+		}
+
+		// Must check needsSync again here because a VMI can be created or
+		// deleted in the startStop function which impacts how we process
+		// hotplugged volumes
+		if c.needsSync(key) && syncErr == nil {
+
+			err = c.handleVolumeRequests(vm, vmi)
+			if err != nil {
+				syncErr = &syncErrorImpl{fmt.Errorf("Error encountered while handling volume hotplug requests: %v", err), HotPlugVolumeErrorReason}
+			}
+		}
+		virtControllerVMWorkQueueTracer.StepTrace(key, "sync", trace.Field{Key: "VM Name", Value: vm.Name})
```

Comment:
Now that this part was moved to a function, the trace step name ("sync") should be the function name.